### PR TITLE
RT-1.12 bgp_always_compare_med

### PR DESCRIPTION
**Pull request by cprabha**
Hi,

Please check new automation for RT-1.12 BGP always compare MED.
This will resolve https://github.com/openconfig/featureprofiles/issues/1556.

Thanks,
Prabha

### CODE REVIEW

File: feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+		t.Logf("Waiting for BGP neighbor to establish...")
+		_, ok := gnmi.Watch(t, dut, nbrPath.SessionState().State(), time.Minute, func(val *ygnmi.Value[oc.E_Bgp_Neighbor_SessionState]) bool {
+			state, ok := val.Val()
+			return ok && state == oc.Bgp_Neighbor_SessionState_ESTABLISHED
+		}).Await(t)
+		if !ok {
+			fptest.LogQuery(t, "BGP reported state", nbrPath.State(), gnmi.Get(t, dut, nbrPath.State()))
+			t.Fatal("No BGP neighbor formed")
+		}
+		status := gnmi.Get(t, dut, nbrPath.SessionState().State())
```
Comment (greg-dennis) [inline feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:197]: No need to look it up again. You can just use the first return value of the Watch call above

File: feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+		}
+	}
+	return ni_proto
+}
+
+// verifyBgpTelemetry checks that the dut has an established BGP session with reasonable settings.
+func verifyBgpTelemetry(t *testing.T, dut *ondatra.DUTDevice) {
+	var nbrIP = []string{ateSrc.IPv4, ateDst1.IPv4, ateDst2.IPv4}
+	t.Logf("Verifying BGP state.")
+	statePath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp()
```
Comment (greg-dennis) [inline feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:183]: Not sure the name "statePath" reflects what this is. Maybe "bgpPath"?

File: feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+		WithFrameSize(512)
+	return []*ondatra.Flow{flowipv41, flowipv42}
+}
+
+// verifyTraffic confirms that every traffic flow has the expected amount of loss (0% or 100%
+// depending on wantLoss, +- 2%).
+func verifyTraffic(t *testing.T, ate *ondatra.ATEDevice, flowName string, wantLoss bool) {
+	// Compare traffic loss based on wantLoss.
+	lossPct := gnmi.Get(t, ate, gnmi.OC().Flow(flowName).LossPct().State())
+	if !wantLoss {
```
Comment (greg-dennis) [inline feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:272]: better to have bool conditions have positive polarity, i.e.

```
if wantLoss {
	if lossPct < 100-tolerancePct {
		...
} else {
	if lossPct > tolerancePct {
		...
}
```

File: feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+	// Verify Port Status.
+	t.Run("Verify port status on DUT", func(t *testing.T) {
+		t.Log("Verifying port status.")
+		verifyPortsUp(t, dut.Device)
+	})
+
+	t.Run("Verify BGP telemetry", func(t *testing.T) {
+		t.Log("Check BGP parameters.")
+		verifyBgpTelemetry(t, dut)
+		// Verify BGP capabilities
```
Comment (greg-dennis) [inline feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:439]: this comment is not necessary, as it just repeats the self-explanatory method call two lines below

File: feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+}
+
+// TestRemovePrivateAS is to Validate that private AS numbers are stripped
+// before advertisement to the eBGP neighbor.
+func TestAlwaysCompareMED(t *testing.T) {
+	// DUT configurations.
+	t.Logf("Start DUT config load.")
+	dut := ondatra.DUT(t, "dut")
+	ate := ondatra.ATE(t, "ate")
+	d := &oc.Root{}
```
Comment (greg-dennis) [inline feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:398]: d is a little bit of a cryptic variable name, even for Go. Maybe rename it "root" throughout?

**Review by greg-dennis — CHANGES_REQUESTED** [review]
This is a very well-structured test. Some small fixups suggested ...

File: feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 2 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: changed it to bgpPath.

File: feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 3 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: updated as suggested.

File: feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 5 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go:398]: changed to dutOcRoot.

File: feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go (hunk 4 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/bgp_always_compare_med/bgp_always_compare_med_test.go]: Removed comment.

**Comment by LimeHat** [comment]
Could you elaborate on changes from the original test plan specified in #1556 ?
It seems that you've moved from 1 port with 2x EBGP sessions to 3 ports with 2x EBGP + 1x IBGP

**Review by greg-dennis — DISMISSED** [review]
Approving for Go quality

**Review by sachendras — DISMISSED** [review]

**Comment by cprabha** [comment]
> Could you elaborate on changes from the original test plan specified in #1556 ? It seems that you've moved from 1 port with 2x EBGP sessions to 3 ports with 2x EBGP + 1x IBGP

To verify traffic for best route selected based on MED , need one traffic source port, hence added port3.  2 eBGP peers on different ports to verify the traffic on the best path.

**Review by sachendras — APPROVED** [review]

**Pull request merged**